### ISSUE TITLE
RubySpec tests not compatible with IEEE 754 round-nearest-to-even mode

### ISSUE THREAD

**Issue by noahgibbs** [ruby/spec#294]
Right now, MRI's sprintf uses IEEE 754 round-nearest-to-even mode (https://en.wikipedia.org/wiki/IEEE_floating_point#Roundings_to_nearest). We're trying to get Float#round to match that behavior (https://bugs.ruby-lang.org/issues/12548). It breaks RubySpec in a few places.

Looking at tests, some are clearly incompatible with that mode. For instance, from core/float/round_spec.rb around line 70:

```
it "returns rounded values for big values" do
    +2.5e20.round(-20).should   eql( +3 * 10 ** 20  )
```

This should round to (+2 \* 10 *\* 20) in that mode, not (+3 \* 10 *\* 20).

All four test failures below are similar -- they seem to depending on the rounding behavior of various types not following IEEE 754 round-nearest-to-even.

Here's Nobuyoshi Nakada's test output when using a local implementation of that mode:

```
1)
Float#round returns rounded values for big values FAILED
Expected 200000000000000000000
to have same value and type as 300000000000000000000

spec/rubyspec/core/float/round_spec.rb:70:in `block (2 levels) in <top (required)>'
spec/rubyspec/core/float/round_spec.rb:3:in `<top (required)>'

2)
Integer#round returns itself rounded if passed a negative value FAILED
Expected 200
 to have same value and type as 300

spec/rubyspec/core/integer/round_spec.rb:22:in `block (2 levels) in <top (required)>'
spec/rubyspec/core/integer/round_spec.rb:4:in `<top (required)>'

3)
Rational#round with no arguments (precision = 0) returns the truncated value toward the nearest integer FAILED
Expected 0
 to equal 1

spec/rubyspec/shared/rational/round.rb:17:in `block (3 levels) in <top (required)>'
spec/rubyspec/core/rational/round_spec.rb:3:in `<top (required)>'

4)
Time#strftime rounds an offset to the nearest second when formatting with %z FAILED
Expected "+01:01:04"
 to equal "+01:01:05"

spec/rubyspec/core/time/strftime_spec.rb:50:in `block (2 levels) in <top (required)>'
spec/rubyspec/core/time/strftime_spec.rb:7:in `<top (required)>'
```

**Comment by eregon** [comment]
I think those are mostly fine to be changed in ruby/spec for the next version, but the real decision has to happen on MRI's bug tracker whether this is bearable for compatibility.
I could imagine some Ruby software calling round and get surprising results by updating the Ruby version.
OTOH Float#to_s changed many times, but the compatibility did not look like a big concern (maybe because we still need printf-like formatting like `"%.3f" % f` to display floats correctly).

**Comment by noahgibbs** [comment]
Matz says on the linked MRI bug that he believes we should change Float#round to match the behavior of sprintf. He hasn't expressed any versioning preference yet that I know of, though.

**Comment by eregon** [comment]
Right, that was my thinking this could get in 2.4 or later, but unlikely to be backported.
I would be happy to see a patch adapting the specs with `ruby_version_is` guards and testing the new behavior :smiley:

**Comment by noahgibbs** [comment]
Opened PR #322 (https://github.com/ruby/spec/pull/322). Ruby 2.4.0dev hasn't integrated Nobu's patch yet, but I'll comment on the corresponding MRI bug that we have a PR for this.

**Comment by noahgibbs** [comment]
Benoit says RubySpec PR #322 looks good. Just waiting for us to commit round-to-nearest-even mode to MRI so he can merge it. I've used Nobu's branch locally, including to create PR #322, and it works well for me. +1

**Comment by mjago** [comment]
Resolved with PR #322 merge.
